src/silo/query_engine/filter_expressions/and.cpp

```diff
@@ -71,61 +78,84 @@ std::tuple<OperatorVector, OperatorVector> And::compileChildren(
    OperatorVector negated_child_operators;
+   std::vector<std::unique_ptr<operators::Predicate>> predicates;
    for (auto& child : all_child_operators) {
       if (child->type() == operators::FULL) {
          continue;
       }
       if (child->type() == operators::EMPTY) {
          OperatorVector empty;
          empty.emplace_back(std::make_unique<operators::Empty>(database_partition.sequenceCount));
-         return {std::move(empty), OperatorVector()};
+         return {
+            std::move(empty),
+            OperatorVector(),
+            std::vector<std::unique_ptr<operators::Predicate>>{}};
       }
       if (child->type() == operators::INTERSECTION) {
          auto* intersection_child = dynamic_cast<operators::Intersection*>(child.get());
          appendVectorToVector(intersection_child->children, non_negated_child_operators);
          appendVectorToVector(intersection_child->negated_children, negated_child_operators);
       } else if (child->type() == operators::COMPLEMENT) {
          negated_child_operators.emplace_back(child->negate());
+      } else if (child->type() == operators::SELECTION) {
+         auto* selection_child = dynamic_cast<operators::Selection*>(child.get());
+         appendVectorToVector<operators::Predicate>(predicates, selection_child->predicates);
+         if (selection_child->child_operator.has_value()) {
+            all_child_operators.emplace_back(std::move(*selection_child->child_operator));
+         }
       } else {
          non_negated_child_operators.push_back(std::move(child));
       }
    }
-   return {std::move(non_negated_child_operators), std::move(negated_child_operators)};
+   return {
+      std::move(non_negated_child_operators),
+      std::move(negated_child_operators),
+      std::move(predicates)};
 }
 
-std::unique_ptr<operators::Operator> And::compile(
+std::unique_ptr<Operator> And::compile(
    const Database& database,
    const DatabasePartition& database_partition,
    AmbiguityMode mode
 ) const {
-   OperatorVector non_negated_child_operators;
-   OperatorVector negated_child_operators;
-   std::tie(non_negated_child_operators, negated_child_operators) =
+   auto [non_negated_child_operators, negated_child_operators, predicates] =
       compileChildren(database, database_partition, mode);
 
+   std::optional<std::unique_ptr<Operator>> index_arithmetic_operator;
    if (non_negated_child_operators.empty() && negated_child_operators.empty()) {
-      return std::make_unique<operators::Full>(database_partition.sequenceCount);
-   }
-   if (non_negated_child_operators.size() == 1 && negated_child_operators.empty()) {
-      return std::move(non_negated_child_operators[0]);
-   }
-   if (negated_child_operators.size() == 1 && non_negated_child_operators.empty()) {
-      return std::make_unique<operators::Complement>(
+      if (predicates.empty()) {
+         return std::make_unique<operators::Full>(database_partition.sequenceCount);
+      }
```

Comment:
```suggestion
      if (predicates.empty()) {
         return std::make_unique<operators::Full>(database_partition.sequenceCount);
      }
      return std::make_unique<operators::Selection>(
         std::move(predicates), database_partition.sequenceCount
      );
```
Then we can remove the `if (index_arithmetic_operator == std::nullopt)` below, since it's only possible to reach that `if` via this branch.

Reply:
Then we can make the `index_arithmetic_operator` a non-optional, since it's always assigned.